16.0 mig customer team manager

Honeyxilia opened the pull request.

fcayre committed:
Make it possible for customers to manage their company's users

B2B customer users can now create, update and deactivate partners as soon as:
- they belong to a customer "admin" group
- those partners are in the same company than them

The customer admins can create a team hierarchy in their company and
attribute them to collegues.

They also can grant access to the portal to their collegues, which
requires to enter their password for the sake of security.

The design choice to make customers directly manipulate partners was
made compared to:
- using a dedicated model with fields synchronized with partners to
  avoid a difficult management of the life cycle of such entities
  compared to partners (when should we create/ unlink them?) and the
  data migration
- using a dedicated entity with delegation inheritance, for the same
  data migration difficulty

The main drawbacks are:
- security: access to other partners is a risk mitigated by the unit
  tests and the CI
- performance: added rules might heart performance, but this is limited
  to portal users and should not heart in the short and medium terms

fcayre committed:
i18n update and fr t10n

fcayre committed:
Make massive import of partners possible by customer

- make ir.translation readable by customer admin (required by import for boolean from str converter)
- use sudo in _load_records_create after security checks (like in create) to allow other modules to
  complete the updated values automatically (parent > child partner sync)

fcayre committed:
Customize export wizard for customer admins

It is very much simplified:
- hide all widgets but the export format file choice
- the import mode is force to import compatibility

This is made client-side (javascript) by adding a "is_customer_admin"
boolean info in the session client-side object, and by overloading the
ExportData widget to add a dedicated css class on the modal html tag +
toggling the right import mode option using javascript at modal display.

fcayre committed:
Customize the partner import example file

fcayre committed:
Prevent customers from removing their own admin role

Also move the "short" tag option into a dedicated kanban widget. This
widget does not handle clicks anymore.

Preventing customers from removing their own admin role is done using a
"readonly" role computed field (which is given current partner_id as a
context). The value of this field is fetched in the widget and used in
coordination with the global edition or view mode to display the removal
icon and to prevent keyboard removal (or not).

fcayre committed:
DE t10n

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Remove the assumption parent_id is not sent when a customer manager creates contacts

... so instead of checking parent_id is set to the permitted value (=the
company of current user), override its value instead.

This is far more reliable and future proof, as the parent_id field may
be sent with False as a value instead of not sent at all depending on
view configuration.

This impacts the contact import feature, as we also need to override
the parent_id field's value when updating a contact, which was unless
before.

fcayre committed:
Refactor tests: explicit the user who performs the import tests

fcayre committed:
Rename test data file

The new name explains better the content of the data file and the difference
with the other data file in the same directory.

fcayre committed:
[tests] Improve test coverage

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Fix crash on company merge when resulting company has no customer admin

fcayre committed:
Show only companies for parent_id field in customer team manager

fcayre committed:
Do not crash when admin customer grants access to invalid email partners

... but display the invalid email partner name list in the wizard.

fcayre committed:
customer_team_manager i18n update and FR + DE t10n completion

fcayre committed:
Post-merge version updates

github-grap-bot committed:
[UPD] README.rst

Honeyxilia committed:
[IMP] customer_team_manager: black, isort, prettier

Honeyxilia committed:
[MIG] customer_team_manager: Migration to 16.0

Honeyxilia committed:
Removed depreciated returns decorator

Honeyxilia committed:
[test] Changed cache invalidation method

To invalidate the cache of the given test user record, since invalidate_cache() is depreciated, we're using invalidate_recordset() to invalidate the cache of the given record (https://github.com/odoo/odoo/commit/32bc28aa665c479e31a40afffca2f03f6f2cc805 )

Honeyxilia committed:
Added batch creation for create() methods

Honeyxilia committed:
Moved asset declaration

Honeyxilia committed:
Replaced old image_medium field call for avatar_128

Starting from Odoo 15.0, the old avatar system has been replaced with a new mixin named avatar.mixin (odoo/addons/base/models/avatar_mixin.py) allowing to retrieve avatars from 128px to 1920px
Since the old image_medium variable used a 128x128 pixels wide image, I used the avatar_128 attribute, following the partner views in `odoo/odoo/addons/base/views/res_partner_views.xml`

Honeyxilia committed:
Removed security group for deleted model

The "ir.translations" model has been deleted, so we have removed the group associated with that model.
=> https://github.com/odoo/odoo/pull/97692

Honeyxilia committed:
Added recursive attribute to full_name

Honeyxilia committed:
Changed customer boolean with replacing customer_rank field

The `customer` boolean attribue has been removed in Odoo 13.0, along with a prioritization of search sorting using the context
=> https://github.com/odoo/odoo/commit/8766f388da9b642ba4433d8d5d670847b472cff8

The replacing behavior is to use a new integer field (inserted from the account module), `customer_rank` :
- This value's propose is to sort customers and suppliers by their numbers of (respectively) sales and purchases
- If the value is higher than 1, the partner is considered a customer (the value is set to one by default if a partner is created through a "Customer" view (communicated through the 'res_partner_search_mode' context)
=> https://github.com/odoo/odoo/commit/46e48055edf1141f7b49989e6e3e9faf246fefd0

Honeyxilia committed:
Added titles to i tags used for icons

This is mainly done for accessibility reasons (description for screen readers)

Honeyxilia committed:
Replaced QWeb view render method

The previous method to render QWeb views was a method included in ir.ui.view.
However, the render of QWeb Views has now two possible methods - _render and render ( https://www.odoo.com/documentation/16.0/fr/developer/reference/frontend/qweb.html#view-based ) - as _render seems to be the prefered rendering method, I used it here.

Honeyxilia committed:
Added new argument to _check_credentials

_check_credentials now takes two arguments : the password and user_agent_env, a dict 'describing any relevant environnement attributes'.
The "interactive" value seems to be about whether the user is interacting with an interface, which is the case here (using the web client)

Honeyxilia committed:
[test] Replaced SavepointCase class by TransactionCase

Honeyxilia committed:
Fixed portal access grant/revoke actions

The used portal actions are declared in the 'portal.wizard.user' model (odoo/addons/portal/wizard/portal_wizard.py)

For the revoke action, since there's a dedicated revoke action in the wizard, we're using it here - however, this adds a couple of changes to the action :
- Revoking deletes the user signup_token, so if the user had never connected to the portal, it is now impossible to sign up.
- Revoking archives the user
- As well as removing the 'portal' group from the user, it also adds the 'public' group
=> 'public' and 'portal' group are concurent - however, customer role groups count as 'portal' groups (since they inherit 'base.group_portal' through the `implied_groups_id` field) - so, we remove the customer role groups before calling the revoke method

Honeyxilia committed:
[test] Added invalidation of cache in tests

When checking the readonly and state values, the first computed value is stored in the cache and is then reused for following tests.
So, we invalidate the cache of the role to force it to recompute the fields and get a data with the new context.

Honeyxilia committed:
[test] Replaced setUp method by setUpClass

Honeyxilia committed:
[test] Fix test to match new login redirect behavior

Beforehand, in 12.0, the response returned a JS code to redirect the user.
Now, the response is a proper HTTP 303 response.

Honeyxilia committed:
Added security rules for Portal Access Wizard

- We grant access to the portal wizard model for customer admins, sale managers but also internal users and portal users because the grant_portal_access method has a built-in allowed group detection.

Honeyxilia committed:
Replaced fields_view_get by get_views

- `fields_view_get` is depreciated from Odoo 16.0 onward, replaced by the methods get_view and get_views - however, only get_views has the functionnality to retrieve the actions in the toolbox.
- We filter the actions in the `form` and `list` views since they're the only `res.partner` views accessible to customer admins
- The toolbox actions in `fields_view_get`/`get_views` no longer contain the xmlid attribute - so, we use the `id` attribute and compare it to the ids of the authorized actions.
Note: the toolbox action ids are from ir.actions.actions, whereas the authorized action is ir.actions.act_window - however, the id of ir_act_window in the postgreSQL table uses the id iterator from ir_actions (irnextval('ir_actions_id_seq'::regclass)), so we can assume that the id of the action of the ir.actions.actions record and the ir.actions.act_window is the same, and so the comparaison is valid.

Honeyxilia committed:
[test] Removed useless statement

Honeyxilia committed:
[test] Changed variables to match more recent implementations

Honeyxilia committed:
Added customer admin attributes skip if the env. superuser tag is positive

In the res.partner write method in the base model, there's a _fields_sync method, which writes "commercial fields" (returned by the res.partner method _commercial_fields) if the parent_id value is changed, as a superuser (with the sudo() method).
Since the modified fields in _commercial_fields are not directly modifiable by a customer admin, instead of adding these commercial fields to the allowed customer admin attributes, we instead skip the check if the superuser environnement variable is False before checking.

Honeyxilia committed:
[test] Fix import_csv method to match name changes in base_import

Honeyxilia committed:
Added permissions for customer admins for base_import.import

Honeyxilia committed:
[test] Changed string check for new error message

Honeyxilia committed:
[test] Change env. user after creation of test partner

Honeyxilia committed:
[test] Changed name from UserError to args[0]

Honeyxilia committed:
Added sudo() for read operation on ir.module.category

When a user in the group_customer_admin tries to fetch the groups, it fails as the domain reads the `category_id` of `res.group`, a `ir.module.category` record which can only be read by internal users.
So, using the sudo() method permits the user to temporarily read this field to perform the action.

Honeyxilia committed:
Reimplemented customer roles tags fields with Many2ManyTagsFields

- The previous widget was based on the `FieldMany2ManyTags` (based on `AbstractField`), which is incompatible with the new registry (for exemple, it handles the fields of a record differently) - as such, we rewrote the widget by extending Many2ManyTagsField (based on the class `Component`).
- Some SCSS classes have changed - we modified our CSS stylesheet to replicate the 12.0 look.

Honeyxilia committed:
Added base_fontawesome dependency

The FontAwesome icons`fa-network-wired` and `fa-coins` are used in this model, and are declared in server-tools/base_fontawesome.
In the production software, the icons are correctly shown because commown_self_troubleshooting also requires it

Honeyxilia committed:
Moved group_customer_admin role to customer_manager_base

This role was intended for customers to manage any resource we offer them, which could be offered in child modules - however, the role was declared in a child module, `customer_team_manager`.
So, we move the role here for better future management.

Honeyxilia committed:
Moved root menuitem to customer_manager_base

customer_manager_base could be expanded beyond the team management functionnality. So, moving the root menuitem allows other menu items to link to it for other features.

Honeyxilia committed:
Added rule to restrict menus accessible to customer admins

With the access to the web app for portal users, we restrict the menus they can access to menu_customer_team_manager_root and their children (in the event we decide to add more management features).

Honeyxilia committed:
Reimplemented export dialog simplification for new ExportDataDialog component

The Export action no longer uses the modal Widget `DataExport`, but a new Componant named `ExportDataDialog`, bringing different changes :
* We removed a couple of classes that force a width and a height attributes on the left panel, which takes the left part of the dialog even when its display is set to None - removing it allows for the Export File Format input to be aligned on the left side of the dialog.
* The template structure has changed, so to hide everything except the export file format, we've had to change the CSS class to replicate the previous look.
* The "import-compatible" radio selection has been changed to a Checkbox - it's unchecked by default, so we click it to have it checked.

Notes :
- Currently, we check the import-compatible checkbox by fetching in the DOM the actual input - this is because in the web.ExportDataDialog template, it's not possible to directly select the `input` element since it's comprised in the CheckBox component (which can't have a `t-ref` attribute itself).
- It's possible to directly set the value of the CheckBox to `true` in the XML template using XPath - however, it then makes it impossible to uncheck it as it stays on true - the reason being that the Dialog box is reloaded each time we click it, and the overwriting of the CheckBox value through our template always applies.

Honeyxilia committed:
Added export-group inheritance to customer admin group

Starting from Odoo 14.0, users have to be included in the `base.group_allow_export` to be able to export data, and to be able to see the Export action in the Actions menu.
Since customer admins should be allowed to export customer data, we grant them access to this group.

This group is not limited to internal users so we can share it with portal users, like the customer admins.

Honeyxilia committed:
Edit customer_manager_base README + icon

Honeyxilia committed:
Filter mail-related navbar items for customer admins

The mail menuitems try to access records only accessible for internal users, which provoke an Access Error for customer admins, who are portal users.
Filtering out these items prevents the access attempt.
This also filters out unused items by customer admins users, decluttering the interface.

Honeyxilia committed:
Added /web route override in customer_manager_base

In the `web` and `portal` modules, the `/web` route controllers contain conditions checking whether the user trying to access the `/web` domain is an internal user, then either redirects or forbids access.
This is problematic since this module's functionnality relies on giving a limited access to the `/web` side of the Odoo app, to be able to perform actions on users limited to their company.
So, we replicated the code in `odoo/addons/web/controllers/web.py` but added a check for customer admins to allow access to /web

NOTE: The way permissions are set, customer admins should only be able to access the records linked to the customer_manager_base module and its children, but this is still granting access to the backend of the Odoo app, so model access rights need to be limited and justified.

Honeyxilia committed:
Moved IrHttp override in customer_manager_base

Since this model is concerned with setting a session variable linked to the `group_customer_admin` group, and this group has been moved to the customer_manager_base module, we moved this model to the corresponding module.

NOTE : In the moved test, we only add the `group_customer_admin` group but no company, since in the current iteration of the test, a company is not required to obtain the `group_customer_admin` group.
A feature which could be added is an override of the `res.users` model that checks whether the user is B2B before being able to get that role - however, this is not the goal of this commit.

Honeyxilia committed:
Modify session_info to fetch `/web` menuitems for customer admins.

In the `web` module, the menuitems are only loaded in the session info properties if the current user is an internal user.
So, we replicate the code in `odoo/addons/web/models/ir_http.py`, and performed the menuitem fetch if the current user is a customer admin.

fcayre commented:
Hello Lena, en lisant la doc de lxml l'autre fois  je suis tombé sur la méthode "form_values" qu'on pourrait avantageusement utiliser dans customer_manager_base/tests/test_ir_http.py dans la méthode get_form où il y a une ligne non couverte (sur les select).

Honeyxilia commented:
> Hello Lena, en lisant la doc de lxml l'autre fois je suis tombé sur la méthode "form_values" qu'on pourrait avantageusement utiliser dans customer_manager_base/tests/test_ir_http.py dans la méthode get_form où il y a une ligne non couverte (sur les select).

Merci!!
Je vais regarder cela, puis si ça peut fonctionner, je l'intègrerai c:
ce sera sans doute poussé après l'intégration des tests de contrôleur

Edit : c'est bon, je l'ai inséré dans le test c:

Honeyxilia committed:
[test] Added test for session expiry

In the `web/controllers/home.py` controller (which we replicated), we check the validity of the session with the `check_session` method, which compares the expected session_token with the given session_token if they match.
Modifying directly the session_token is not possible : before the controller call, the session validity is checked, so modifying the token already invalidates the session, which gives an anonymous session.

To circumvent this, we created a mock which returns True on the first call (before the controller, to not invalidate the session) but False on subsequent calls (normally, only the controller call), so we can easily match the condition.*

NOTE: For the calls list we use to count the number of calls, we make use of a dangerous default value in order to broaden the scope of the calls list to both the test method and the test function, so the list object can be modified and have its changes kept in memory in both scopes.

Honeyxilia committed:
[test] Added tests for partner with public user

Honeyxilia committed:
[test] Simplified form values extraction

Honeyxilia committed:
[test] Improve coverage: regrant access to portal + comments

Honeyxilia committed:
[test] Improve coverage: Checking wizard info from a sale manager's point of view

Honeyxilia committed:
[test] Improved coverage: using create_by_form with full model name

In create_by_form, a check has been impremented to concatenate the name of the module in front of the model name if a short name is passed.
Using the full model name here allows for coverage both for the short version and the full version.

Honeyxilia committed:
[test] Improved coverage: Checking company_name write authorization

In the write overload of the res.partner model, allowed users are only able to change the value of the company_name field if it's to set it to False.

Honeyxilia committed:
[test] Improved coverage: Check unlink override for non-B2C partners

Honeyxilia committed:
[test] Imp. coverage: Check get_views from non-admin's point of view

9184Lucky commented on customer_team_manager/models/wizard_portal_access.py:
```diff
@@ -87,12 +87,13 @@ def grant_portal_access(self):
-    @api.model
-    def create(self, vals):
+    @api.model_create_multi
+    def create(self, vals_list):
```
Vu que on parle ici d'un wizard, je ne suis pas sûr que ce soit pertinent de faire une méthode de création en batch (j'ai l'impression que c'est un cas que on ne peut pas rencontrer)... Peut être faut il simplement désactiver ce warning si c'est possible. A discuter ... :)

Honeyxilia replied:
En effet, c'était principalement pour l'avertissement - cependant, même si on n'utilise pas la fonctionnalité (à cause de sa pertinence), à part si l'on veut empêcher explicitement la création multiple de wizard, je ne considère pas cela comme un aspect négatif (mais c'est mon opinion)

fcayre replied:
Je suis d'accord avec Lena, maintenant que le travail est fait ça ne sert à rien de le défaire. Comme le dit Luc le faire lorsque ce n'est pas pertinent n'est pas un effort rentable. En passant vous deviez vous en douter, mais ce n'est qu'une question de performance (donc ce n'est jamais strictement nécessaire, mais c'est une bonne pratique de le faire !).

9184Lucky commented on customer_team_manager/models/res_partner.py:
```diff
@@ -126,9 +126,11 @@ def copy_data(self, default=None):
+        Raises AccessError if vals is not a superuser and has unallowed attributes.
         """
+        if self.env.su:
+            return
```
Hello, je ne comprends pas trop l'idée ici. Ici non seulement on saute les values checks (pourquoi pas) mais surtour on fait aussi sauter certains mécanismes comme le fait d'ajouter les roles si c'est le seul utilisateur de la company... Du coup je ne comprends pas trop la raison et surtout, de l'éxterieur j'ai l'impression que on ne se contente pas de porter une fonctionalité en v16 mais que on modifie son comportement....
Bref, j'ai peut être loupé un truc mais je veux bien un éclairage sur la question :)

Honeyxilia replied:
Ce check est fait, car dans le modèle res.partner, il y a une opération de synchronisation de champs "commerciaux" (comme `vat`, `company_registry`, ...) qui sont écrits avec la méthode `write` en utilisant `sudo()` pour s'accorder des droits superutilisateur - or, dans notre surcharge de `write`, on n'autorise pas particulièrement ces champs pour les admins clients, qui n'utiliseront pas cette méthode avec `sudo()`, donc ça n'ajoute pas des permissions en plus pour les admins client·es (selon moi)

-> Je peux améliorer le message du commit pour préciser cela c:

Cet ajout n'empêche cependant pas l'ajout de rôles pour le seul utilisateur de la compagnie.

9184Lucky replied:
Hello ! My bad j'avais du lire avec le code "replié" et j'ai pensé que y'avait dans cette fonction tout un tas de ligne qui n'y sont pas !!! Du coup une partie de mon comentaire n'est pas pertinent déso !!

Par contre si le but est de désactiver le check pour certains cas que l'on connait, plutôt que de passer le test sur la présence de sudo j'aurais utilisé une variable de que tu passe dans le contexte et dont tu test la présence en début de code. C'est moins générique et permet de désactiver le test explicitement quand tu le souhaite. 
Exemple: avec une variable skip_customer_allowed_attribut_check que tu test ici "if self.env.context.get("skip_customer_allowed_attribut_check")" et que tu met dans le contexte quand tu en as besoin avec with_context()

fcayre replied:
Il faudrait commencer par rappeler (dans un commentaire dans le code) pourquoi ces lignes sont nécessaires : dans quel contexte ? Je pense en fait qu'on ne devrait tout simplement pas appeler cette fonction quand on est su.

Honeyxilia replied:
Cette méthode est appelée en `su`, car lors de l'opération d'écriture ou de création, une sous opération de `res.partner` si `parent_id` est modifié effectue une opération d'écriture sur des champs protégés retournés par `_commercial_fields`, et afin de faire cette écriture, elle utilise `write` avec `sudo`
https://github.com/odoo/odoo/blob/3f64c4dc6de47beebffbde424b92fff0e0c22fd3/odoo/addons/base/models/res_partner.py#L635

Je vais modifier le commit afin que les conditions fassent en sortent que la méthode ne soit pas appelée du tout si `su` est positif, et modifier le commentaire de commit afin de plus développer

9184Lucky commented on customer_team_manager/tests/test_res_partner.py:
```diff
@@ -250,7 +250,7 @@ def test_grant_and_revoke_portal_access(self):
         self.assertEqual(empl.portal_status, "never_connected")
 
-        self.simulate_user_login(empl.sudo().user_ids)
+        self.simulate_user_login(empl)
```
C'est un détail mais je ne comprends pas ce que viens faire cette ligne dans ce commit (ni du coup la raison de la modif même si j'immagine que c'est par ce que ce n'était plus utile)

Honeyxilia replied:
La différence ici, c'est qu'on ne donne pas l'utilisateur lié à un partenaire en tant qu'argument, mais le partenaire en soi. En effet, dans le même commit, on modifie simulate_user_login pour qu'un partenaire soit pris en charge.

La raison à cela est que la variable portal_status était gardée en cache avant et après cet appel, ce qui fait que malgré la "connexion", le statut reste sur `never_connected` et passe pas à `already_connected`.
Ainsi, en appelant la méthode, j'applique la même opération sur l'utilisateur, mais je vide le cache du partenaire afin de m'assurer que la valeur de `portal_status` soit bien recalculée

fcayre commented on customer_team_manager/models/res_partner.py:
```diff
@@ -228,10 +228,21 @@ def unlink(self):
+        model = self.env["portal.wizard"].with_context(active_ids=[self.id])
+        wizard = model.sudo().create({})
+
+        portal_users = wizard.user_ids.filtered(
```
Ceci effectue une boucle sur wizard.user_ids. Or il y en a une autre juste en dessous donc je pense qu'il vaudrait mieux boucler sur wizard.user_ids et faire le if dedans.

Honeyxilia replied:
Appliqué

fcayre commented on customer_team_manager/models/res_partner.py:
```diff
@@ -228,10 +228,21 @@ def unlink(self):
+        for portal_user in portal_users:
+            # Remove all customer role groups before using the revoke actions,
+            # since they count as "portal" roles and since block the "public" role writing.
+            role_model = self.env["customer_team_manager.customer_role"]
```
À sortir de la boucle pour améliorer les perfs.

Honeyxilia replied:
Appliqué c:

fcayre commented on customer_team_manager/models/wizard_portal_access.py:
```diff
@@ -60,26 +60,26 @@ def mail_ext(email):
+            lambda user: user.partner_id in partners and not user.is_portal
+        )
+
+        if non_portal_users:
```
Le if est inutile ici (si la liste est vide, le for ne fera rien) : on peut le supprimer.

Honeyxilia replied:
Appliqué

fcayre commented on customer_team_manager/security/ir.model.access.csv:
```diff
@@ -9,3 +9,7 @@ acl_customer_role_users,acl_customer_team_manager_role_users,model_customer_team
+acl_customer_team_manager_portal_access_wizard_group_portal,acl_customer_team_manager_portal_access_wizard_group_portal,model_customer_team_manager_portal_access_wizard,base.group_portal,1,0,1,0
+acl_customer_team_manager_portal_access_wizard_group_system,acl_customer_team_manager_portal_access_wizard_group_system,model_customer_team_manager_portal_access_wizard,base.group_system,1,0,1,0
+acl_customer_team_manager_portal_access_wizard_admin_customer,acl_customer_team_manager_portal_access_wizard_admin_customer,model_customer_team_manager_portal_access_wizard,customer_team_manager.group_customer_admin,1,0,1,0
+acl_customer_team_manager_portal_access_wizard_sale_manager,acl_customer_team_manager_portal_access_wizard_sale_manager,model_customer_team_manager_portal_access_wizard,sales_team.group_sale_manager,1,0,1,0
```
Je ne suis pas sûr de comprendre : on utilise justement sudo dès qu'on crée un portal wizard, donc les utilisateurs en question n'ont pas besoin de ces droits si ?

Honeyxilia replied:
C'était principalement parce que dans le log, il soulève le fait qu'il n'y ait pas de permissions définies pour le modèle `portal_access_wizard`

Egalement, si tu parle du `sudo()` à la ligne 73 du fichier `customer_team_customer/models/wizard_portal_access.py`, il permet de créer un `portal.wizard` et non un `customer_team_manager.portal_access_wizard` - or, pour lancer l'action d'autorisation d'accès au portal, il faut créer ce dernier modèle, et sans ces règles, il n'est pas possible d'en créer un depuis l'interface

fcayre commented on customer_team_manager/views/customer_team_manager.xml:
```diff
@@ -69,18 +69,20 @@
                   </li>
-                  <li><i class="fa fa-users" /> <field name="team" /></li>
-                  <li><i class="fa fa-phone" /> <field
+                  <li><i class="fa fa-users" title="Team" /> <field name="team" /></li>
```
C'est probablement cool d'ajouter un title aux tags `<i>` mais il faudrait dire pourquoi dans le message de commit.

Honeyxilia replied:
Appliqué

fcayre commented on customer_manager_base/tests/test_ir_http.py:
```diff
@@ -83,9 +83,8 @@ def test_session_info_is_customer_admin_true(self):
             env = self.env(test_cursor)
             env.ref("customer_team_manager.customer_role_admin")
```
Cette ligne est inutile et devrait pouvoir être effacée.

Honeyxilia replied:
elle est déjà supprimée dans un commit futur

fcayre commented on customer_manager_base/__manifest__.py:
```diff
@@ -18,6 +18,7 @@
     ],
     "data": [
         "security/groups.xml",
+        "security/ir.model.access.csv",
```
Typos dans le message de commit : 
- proprieties => properties
- replicated => replicate (ou perform => perfomed pour la concordance des temps, mais je préfère faire simple)

Honeyxilia replied:
Fait

fcayre commented on customer_manager_base/models/ir_http.py:
```diff
@@ -7,7 +12,50 @@ class IrHttp(models.AbstractModel):
     def session_info(self):
         "Complete session info with `is_in_group_user` for use in js code (eg. widgets)"
```
La docstring n'est plus à jour.

Honeyxilia replied:
Mis à jour c:

fcayre commented on customer_manager_base/README.rst:
```diff
@@ -13,7 +13,7 @@ Customer manager base
     :target: https://odoo-community.org/page/development-status
     :alt: Alpha
-.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
+.. |badge2| image:: https://img.shields.io/badge/license-AGPL--3-blue.png
```
Je n'ai pas compris pourquoi avoir changé l'icône. Elle est désormais identique à celle de customer_team_manager du coup non (ce qui n'aide pas à les distinguer) ? Après j'avoue ne pas avoir de politique bien claire sur les icônes...

Honeyxilia replied:
Etant donné qu'on a déplacé le menuitem d'accès au module de `customer_team_manager` à `customer_manager_base`, il n'utilisait plus l'icone -> De cette manière, l'icone est toujours utilisée

fcayre commented on customer_team_manager/__manifest__.py:
```diff
@@ -36,7 +36,7 @@
             "/customer_team_manager/static/src/css/export.css",
             "/customer_team_manager/static/src/scss/field_widget.scss",
-            "/customer_team_manager/static/src/js/field_widget.js",
+            "/customer_team_manager/static/src/js/field_widget.esm.js",
```
Le message de commit est un peu trop verbeux et contient au moins une typo (to better => to improve).

Je suis pour des messages de commit bien explicites mais peut-être un peu plus efficaces.

fcayre commented on customer_team_manager/__manifest__.py:
```diff
@@ -34,10 +34,11 @@
     "assets": {
         "web.assets_backend": [
             "/customer_team_manager/static/src/xml/field_widget_template.xml",
+            "/customer_team_manager/static/src/xml/export_data_dialog.xml",
```
Typos dans le message de commit : 
- componant => component

Le message de commit est trop verbeux ici : il n'a pas vocation à servir de doc à l'API de Component.
Tes notes sont à mon avis la partie la plus pertinente, qui expliquent tes choix de conception et permettront de faire gagner beaucoup de temps à une personne qui se demanderait pourquoi c'est fait comme ça.

Honeyxilia replied:
Je pense que certaines des notes de changement peuvent être pertinantes afin d'expliquer les changements, mais j'ai retiré la ligne sur les t-ref et celle sur patch/onMounted c:

fcayre commented on customer_team_manager/static/src/js/customer_admin_navbar.esm.js:
```diff
@@ -0,0 +1,23 @@
+     */
+    get systrayItems() {
+        var navbar_items = this._super();
+        if (session.is_customer_admin) {
```
On ne peut pas plutôt transformer le test en "n'est pas un utilisateur interne" ? (pas sûr qu'on ait l'info). Ce serait plus logique et plus "future proof".

En passant, typo dans le message de commit : "accessible for" => "accessible by".

Honeyxilia replied:
Ce test est basé sur la variable de session établie dans `customer_manager_base/models/ir_http.py` -> On peut créer une nouvelle variable de session si l'on souhaite en effet

Honeyxilia replied:
Ainsi, j'ai appliqué cela c:
J'en profite également pour déplacer ce fichier de customer_team_manager à customer_manager_base, car c'est là où l'on récupère également les menus

fcayre commented on customer_team_manager/tests/test_res_partner.py:
```diff
@@ -146,6 +146,21 @@ def test_write_email(self):
+        company_as_admin = self.customer_company.with_user(self.customer_user_admin)
+
+        company_as_admin.write({"company_name": False})
+        self.assertEqual(
```
Utiliser plutôt self.assertFalse

Honeyxilia replied:
Appliqué

fcayre commented on customer_team_manager/tests/test_res_partner.py:
```diff
@@ -201,6 +201,13 @@ def test_write_active_false_removes_portal_user(self):
+
+        non_empl = self.create_partner(name="Not Empl", email="not_emp@test.coop")
+        res = non_empl.unlink()
+        self.assertTrue(res)
```
On vérifie ici seulement que unlink ne plante pas non ? Du coup il faut ajouter un commentaire et je pense enlever le assertTrue qui semble dire que si on était en B2B on n'aurait False...

Honeyxilia replied:
Le True assure que l'opération s'est bien passée (voir fichier `odoo/odoo/models.py`) et soulève une erreur (AccessError/UserError) si l'opération échoue - le code qu'on ajoute un autre soulèvement d'erreur (ValidationError) si l'opération `_check_one_customer_admin_at_least` échoue, mais elle ne retourne pas False s'il s'agit un B2B

Cependant, puisque ce n'est pas nécessaire d'être vérifié (une exception se verrait immédiatement), je retire donc cet assert

fcayre commented:
/ocabot merge patch

Honeyxilia committed:
Added optional argument to _action_launch_stock_rule overload

Honeyxilia commented:
/ocabot merge patch

fcayre commented:
/ocabot merge patch
